### PR TITLE
Add tests for generic types/values with the Marshal APIs

### LINKED ISSUE #35193: Add Marshal API test coverage for generics marshalling
#103 added support for marshalling generics, but we don't seem to have test coverage for the Marshal APIs with this:

* Marshal.OffsetOf
* Marshal.SizeOf
* Marshal.PtrToStructure
* Marshal.StructureToPtr
* Others?

E.g. Marshal.OffsetOf already worked on desktop CLR because the runtime didn't do a good job of blocking it (and it also raises a question as to whether the code paths this is taking are quite different that made it so that blocking didn't take place). We don't have tests that the returned numbers are actually good.

We might as well throw in a test for marshalling as UnmanagedType.AsAny because I'm sure WinForms is keen to use that (`/snark`).

I'm making this a .NET 5 issue since this is a new feature.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Adds tests that validate the current behavior of OffsetOf with generic types

Fixes #35193

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/OffsetOfTests.cs [hunk 1]
```diff
@@ -224,6 +224,20 @@ public void OffsetOf_NoLayoutPoint_ThrowsArgumentException()
...
+        {
+            Assert.Equal((IntPtr)4, Marshal.OffsetOf<Point2<int>>(nameof(Point2<int>.y)));
+            Assert.Equal((IntPtr)8, Marshal.OffsetOf<Point2<ulong>>(nameof(Point2<int>.y)));
+            Assert.Equal((IntPtr)4, Marshal.OffsetOf<Point2<float>>(nameof(Point2<int>.y)));
+            Assert.Equal((IntPtr)8, Marshal.OffsetOf<Point2<double>>(nameof(Point2<int>.y)));
+
+            // [COMPAT] Non-blittable generic types with value-type generic arguments are supported in OffsetOf since they've always been allowed
+            // and it likely doesn't meet the bar to break back-compat.
+            Assert.Equal((IntPtr)1, Marshal.OffsetOf<Point2<char>>(nameof(Point2<int>.y)));
+            Assert.Equal((IntPtr)1, Marshal.OffsetOf<Point2<byte>>(nameof(Point2<int>.y)));
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/OffsetOfTests.cs:238]: Can we also add a test case for the non-value-type behaviour?

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/OffsetOfTests.cs:238]: We can add a case for a non-value-type generic, but we can't add a reliable test for an instantiation with a reference type.

Whether or not `Point2<string>` is supported depends on if there's a canonical instantiation (aka is the actual instantiation at runtime `Point2<System.__Canon>` or `Point2<string>`)

**Comment by jkoritzinsky** [comment]
I've added more tests based on the discussion in #35193

**Review by elinor-fung — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
This PR is only test changes and the failures are in unrelated tests. Merging

**Pull request merged**